osu.Game/Screens/Play/Player.cs

```diff
@@ -719,10 +719,16 @@ public bool Restart(bool quickRestart = false)
             // stopping here is to ensure music doesn't become audible after exiting back to PlayerLoader.
             musicController.Stop();
 
-            if (RestartRequested != null)
+            skipExitTransition = quickRestart;
+            PrepareLoaderForRestart?.Invoke(quickRestart);
+
+            if (!this.IsCurrentScreen())
             {
-                skipExitTransition = quickRestart;
-                RestartRequested?.Invoke(quickRestart);
+                // if we're called externally (i.e. from results screen),
+                // use MakeCurrent to exit results screen as well as this player screen
+                // since ValidForResume = false in here
+                Debug.Assert(!ValidForResume);
+                this.MakeCurrent();
```

Comment:
This hurts my head. I guess at least there's an inline comment next to it but it's still a solution I'd classify as convoluted at best.

Can this not be done any simpler? I'd much rather have a duplicated `PerformExit()` call in the original code in the `RestartRequested` block or something than this.

Reply:
@bdach does this read any better to you?

```diff
diff --git a/osu.Game/Screens/Play/Player.cs b/osu.Game/Screens/Play/Player.cs
index 3a0a0613f3..1866ed26ce 100644
--- a/osu.Game/Screens/Play/Player.cs
+++ b/osu.Game/Screens/Play/Player.cs
@@ -646,7 +646,6 @@ protected bool PerformExit(bool skipTransition = false)
             // import current score if possible.
             prepareAndImportScoreAsync();
 
-            // Screen may not be current if a restart has been performed.
             if (this.IsCurrentScreen())
             {
                 skipExitTransition = skipTransition;
@@ -657,6 +656,12 @@ protected bool PerformExit(bool skipTransition = false)
                 // - the pause / fail dialog was requested but couldn't be displayed due to the type or state of this Player instance.
                 this.Exit();
             }
+            else
+            {
+                // May be restarting from results screen.
+                if (this.GetChildScreen() != null)
+                    this.MakeCurrent();
+            }
 
             return true;
         }
@@ -722,16 +727,6 @@ public bool Restart(bool quickRestart = false)
             skipExitTransition = quickRestart;
             PrepareLoaderForRestart?.Invoke(quickRestart);
 
-            if (!this.IsCurrentScreen())
-            {
-                // if we're called externally (i.e. from results screen),
-                // use MakeCurrent to exit results screen as well as this player screen
-                // since ValidForResume = false in here
-                Debug.Assert(!ValidForResume);
-                this.MakeCurrent();
-                return true;
-            }
-
             return PerformExit(quickRestart);
         }
 

```

Reply:
The problem I have is the whole "making a non-valid-for-resume screen current in order to implicitly exit it" shebang. Not the positioning of it in code. So I guess that's a no.

Reply:
An alternative option I can imagine would be to move the `Restart` method to `PlayerLoader` instead and choose between calling `Player.PerformExit()` and `MakeCurrent` based on whether player is current screen. That will solve the hack of making a non-valid-for-resume screen current.

Reply:
> An alternative option I can imagine would be to move the `Restart` method to `PlayerLoader` instead and choose between calling `Player.PerformExit()` and `PlayerLoader.MakeCurrent()` based on whether player is current screen.

Yeah I don't know if I'd consider this better.

Maybe this is fine if there really is no saner alternative, I dunno.

Reply:
I will say that I have attempted my proposal above just to see where it goes and how bad/good would it be, but upon stumbling across [some existing test code](https://github.com/ppy/osu/blob/9140893249037130c9a2bae7bd67be20f8be098a/osu.Game.Tests/Visual/Mods/TestSceneModFailCondition.cs#L21-L26) that relies on the current structure I halted down immediately as totally not worth the effort.

I like peppy's diff as it makes the logic look more appropriate than it is currently, pushed in https://github.com/ppy/osu/pull/30937/commits/fa87df6c6af7879f1bc2e60b276724dddd3c2136.